src/rdkafka_broker.c

```diff
@@ -640,14 +644,43 @@ static int rd_kafka_broker_bufq_timeout_scan (rd_kafka_broker_t *rkb,
 
 		rd_kafka_bufq_deq(rkbq, rkbuf);
 
-                if (now && cnt < log_first_n)
+                if (now && cnt < log_first_n) {
+                        char holbstr[128];
+                        /* Head of line blocking:
+                         * If this is not the first request in queue, but the
+                         * initial first request did not time out,
+                         * it typically means the first request is a
+                         * long-running blocking one, holding up the
+                         * sub-sequent requests.
+                         * In this case log what is likely holding up the
+                         * requests and what caused this request to time out. */
+                        if (holb && holb == TAILQ_FIRST(&rkbq->rkbq_bufs)) {
+                                rd_snprintf(holbstr, sizeof(holbstr),
```

Comment:
nice log message. is the `holb == TAILQ_FIRST(&rkbq->rkbq_bufs` condition superfluous? (always true if `hobl != NULL`)?

Reply:
holb points to the first element in the queue before anything has timed out.
At this point the first element, holb, might have timed out and removed from the queue, so it might no longer be the first element.